lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -248,6 +248,23 @@ defmodule KinesisClient.Stream.Shard.Producer do
             message:
               "Checkpoint failed for #{state.app_name}-#{state.shard_id}: this consumer is not the lease owner"
 
+        {:error, {"ProvisionedThroughputExceededException", _}} ->
+          if attempt > 10 do
+            raise KinesisClient.Error,
+              message:
+                "Checkpoint failed for #{state.app_name}-#{state.shard_id}: provisioned throughput exceeded after 10 attempts"
+          end
+
+          :telemetry.execute(
+            [:kinesis_client, :shard_producer, :checkpoint_error],
+            %{reason: :provisioned_throughput_exceeded, attempt: attempt},
+            telemetry_meta(state)
+          )
+
+          # Retry on failure to update checkpoint after jittery sleep
+          Process.sleep((100 + :rand.uniform() * 100) |> floor())
```

Comment:
I think we might want to use something a little smarter here. I just checked and the DynamoDB docs recommend an exponential backoff when retrying rate limit errors (see https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.RetryAndBackoff)